gstreamer-video/src/video_codec_frame.rs

```diff
@@ -0,0 +1,79 @@
+// Copyright (C) 2017 Thibault Saunier <tsaunier@gnome.org>
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use ffi;
+use glib::translate::*;
+use glib_ffi;
+use gobject_ffi;
+use std::mem;
+use std::ptr;
+use gst;
+use gst::miniobject::MiniObject;
+
+glib_wrapper! {
+    pub struct VideoCodecFrame(Shared<ffi::GstVideoCodecFrame>);
+
+    match fn {
+        ref => |ptr| ffi::gst_video_codec_frame_ref(ptr),
+        unref => |ptr| ffi::gst_video_codec_frame_unref(ptr),
+        get_type => || ffi::gst_video_codec_frame_get_type(),
+    }
+}
+
+impl VideoCodecFrame {
```

Comment:
Subclasses need to be able to modify the fields of this, which is unfortunately not thread-safe (and thus `Send` and `Sync` and not exactly correct)

Reply:
Yeah, should we figure out a proper way to do it right now? Do you have ideas to investigate?

Reply:
For now I would do a manual implementation for the getters/setters.

Not sure what to do about the thread-safety issue, as nothing is going to prevent you from getting a new reference at any time from another thread. Maybe another layer of wrapping would be needed, so that accessing any content of the frame would take the stream lock too.

Reply:
Could also everywhere where you can get a frame, instead return a struct that has a reference to the actual frame and the decoder, and on lock() locks the stream lock and returns a mutex guard that on Deref returns a frame reference

Same for codec state